### ISSUE TITLE
Implement HyperplaneMultiHeadLUT (generalize FastMultiHeadLut anchor-pair sign tests to learned hyperplanes)

### ISSUE THREAD

**Issue by nucstarbot** [anatoli-starostin/spiky#64]
> Spec issue for the `research/hyperplane_ffn` line of work. This is a **specification for a GPU-having agent to implement** — do not merge until reviewed. No implementation is started here.

## Goal

Generalize the index-computation step of `FastMultiHeadLut` (`src/spiky/lutorch/fast_multi_head_lut.py`) from **fixed anchor-pair sign tests** to **fully learned hyperplanes**. New module: **`HyperplaneMultiHeadLUT`**, cloned from `FastMultiHeadLut`.

The old `MultiHeadLut` (`src/spiky/lutorch/multi_head_lut.py`, the `AnchorPairsLookup` + `LProjection` composition) is **deprecated — do not touch it.** All work builds on `FastMultiHeadLut`.

## The idea

In `FastMultiHeadLut`, each of the `NAP` index bits comes from a **fixed anchor pair** `(p1, p2)` chosen at init:

```
bit_i = 1[ x[p1] - x[p2] > 0 ]
```

That is a sign test against the special hyperplane with normal `(e_p1 - e_p2)` through the origin — a comparator restricted to one axis-aligned difference direction, with zero threshold.

**Generalize:** replace each of the `NAP` anchor-pair tests with a general **learned hyperplane**:

```
bit_i = 1[ <w_i, x> + b_i > 0 ]
```

where `w_i` is a trainable weight vector over **all** input components and `b_i` is a trainable bias / threshold.

Pack the `NAP` sign bits **MSB-first** into the table index exactly as before. Everything downstream is structurally unchanged:

- `F.embedding_bag(mode='sum')` reduce over the `tables_per_head` axis,
- both `hard` and `hybrid_smooth` forward modes,
- the always-soft, full-K-softmax surrogate backward.

Only the index-computation (the sign-pack front-end) and the gradients it produces change.

## Design decisions (confirmed)

1. **Bias: YES.** Hyperplanes are affine: `<w_i, x> + b_i > 0`, with a learnable threshold `b_i`.
2. **Weight scope: one set of `NAP` hyperplanes PER TABLE.** This is the natural analog of per-table anchor pairs.
   - Weight tensor shape `[n_tables, NAP, d_model]`, where `n_tables = n_heads * tables_per_head`.
   - Bias tensor shape `[n_tables, NAP]`.
3. **Soft backward — the one nontrivial port (core deliverable).**
   The current soft path treats the forward as a full-K softmax pinned to the chosen index, scatters `grad_x` back to the anchor positions, and learns the two temperatures `T_soft` (per-anchor sign sharpness) and `T_sel` (row selection). In the anchor-pair version the differentiable pre-activation is `d_i = x[p1] - x[p2]`, so the only input gradient is a `±d` scatter at two coordinates.

   In the hyperplane version the differentiable pre-activation is `a_i = <w_i, x> + b_i`, so the gradient must now flow into **all three** of:
   - `x` — via `w_i` (dense over `d_model`, no longer a 2-coordinate scatter),
   - the hyperplane weights `w_i`,
   - the biases `b_i`.

   Adapt the soft path to differentiate through the affine pre-activation `a_i` in place of the pairwise difference `d_i`. Keep the same surrogate structure (full-K softmax over the `2^NAP` rows, pinned to the chosen index, with `T_soft`/`T_sel`). **Treat this as the core implementation task of the issue.**

## Requirements

- **Clone `FastMultiHeadLut`; keep it as efficient as possible.** Preserve all of: fused ops, the `lutorch_cuda` CUDA fast path / compiled fallback for hard eval, `weight_dtype` vs autocast handling, and the weight-grad backend auto-pick (bf16 → sparse-S + bmm, fp32 → `index_add`).
- **The per-table hyperplane projection** is a batched matmul `[B, d_model] × [n_tables, NAP, d_model] → [B, n_tables, NAP]`; implement it efficiently (bmm / einsum, dtype-aware, autocast-friendly). This replaces the cheap `x[a] - x[b]` gather and is the new dominant FLOP cost — size and benchmark it.
- **Preserve both forward modes** (`hard`, `hybrid_smooth`) and the **always-soft backward**, now additionally producing grads for `w_i` and `b_i`.
- **Initialization:** support an init that **reproduces fixed anchor pairs** exactly (`w_i = e_p1 - e_p2`, `b_i = 0`) so the module can be dropped in as a strict generalization and A/B'd against `FastMultiHeadLut`. Also provide a sensible general random init (e.g. small-norm Gaussian rows, zero bias) for the learned-from-scratch case.

## Tests (thorough)

- **`gradcheck` on the soft backward** for `w`, `b`, and `x` (double precision, small shapes) — the numerical vs analytic gradient must match within tolerance for all three.
- **Index-packing correctness:** MSB-first bit ordering identical to `FastMultiHeadLut`; the packed index equals the expected `Σ_i bit_i · 2^(NAP-1-i)`.
- **Parity with `FastMultiHeadLut`:** when hyperplanes are initialized to reproduce fixed anchor pairs (`w_i = e_p1 - e_p2`, `b_i = 0`), the forward output (both `hard` and `hybrid_smooth`) and the `x`/weight/temperature gradients must match `FastMultiHeadLut` to numerical tolerance.
- **Both forward modes** exercised; `hard` ↔ `hybrid_smooth` runtime flip works.
- **dtype coverage:** fp32 weights + bf16 autocast, and bf16 storage, on CUDA when available; CPU fallback path.
- **A tiny end-to-end train step** (overfit a small batch) to confirm loss decreases and `w`/`b`/temps all receive nonzero grads.

## Non-goals / out of scope

- Do not modify or delete the deprecated `MultiHeadLut`.
- No changes to LUTGPT model wiring in this issue — this is the standalone primitive plus tests. Integration (e.g. as an FFN / `out_proj` replacement) is a follow-up.
- No claim that learned hyperplanes beat fixed anchors on val bpb yet — that's the experiment the primitive enables, not part of this issue.

## Acceptance criteria

- `HyperplaneMultiHeadLUT` lives alongside `FastMultiHeadLut`, same public forward signature (`x: [B, d_model] → [B, n_heads, n_outputs]`), plus exposed `w`/`b` parameters.
- All tests above pass.
- Anchor-pair-equivalent init reproduces `FastMultiHeadLut` bit-for-bit (within tolerance).
- A short benchmark note comparing the new hyperplane projection's forward/backward cost vs the anchor-pair front-end at LUTGPT shapes.

---

**Branch / reference:** `research/hyperplane_ffn`. Base the module on `src/spiky/lutorch/fast_multi_head_lut.py`; see `claude/thesis.md` and `doc/lutorch/lutgpt_research_report.pdf` (Section 2) for the soft-backward math.
